### PR TITLE
[TF FE] Support complex tensors for ScatterNd operations

### PR DESCRIPTION
**Pull request by MonalSD**
### Details:
- *Extended loader ScatterNd by propagating ComplexTypeMark from input to output and to represent output complex type tensor as a floating-point type tensor with an auxiliary dimension that concatenates real and imaginary parts of complex tensor.*
- *Wrapped the complex result with ComplexTypeMark and returned the result*

### Tickets:
 - #23240

### CODE REVIEW

**Comment by MonalSD** [comment]
Hi @rkazants please do check the changes. Thanks

File: src/frontends/tensorflow_common/src/op/scatter_nd.cpp [hunk 1]
```diff
@@ -15,15 +15,34 @@ namespace frontend {
 namespace tensorflow {
 namespace op {
 OutputVector translate_scatter_nd_op(const NodeContext& node) {
-    default_op_checks(node, 3, {"ScatterNd", "SCATTER_ND"});
+    default_op_checks(node, 3, {"ScatterNd", "SCATTER_ND"}, true);
     auto input_indices = node.get_input(0);
     auto updates = node.get_input(1);
     auto shape = node.get_input(2);
 
+    auto complex_type_mark_input_indices = as_type_ptr<ComplexTypeMark>(input_indices.get_node_shared_ptr());
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/scatter_nd.cpp:23]: indices can not be of complex type

File: tests/layer_tests/tensorflow_tests/test_tf_ScatterND.py [hunk 2]
```diff
@@ -73,3 +73,54 @@ def test_tf_scatter_nd(self, params, ie_device, precision, ir_version, temp_dir,
...
+        dict(x_shape=[4, 4, 4], indices=[[0], [1], [3]], updates=[
+            [[5j, 1j, 5j, 13j], [8j, 6j, 6j, 8j], [7j, 0j, 0j, 7j], [8j, 8j, 8j, 8j]],
+            [[0j, 0j, 0j, 0j], [1j, 2j, 3j, 4j], [5j, 6j, 7j, 8j], [9j, 10j, 11j, 12j]],
+            [[5j, 5j, 5j, 5j], [6j, 6j, 6j, 6j], [7j, 7j, 7j, 7j], [8j, 8j, 8j, 8j]]]),
+        dict(x_shape=[2, 2, 2], indices=[[1, 1, 1], [0, 1, 0]], updates=[9j, 6.3j]),
+        pytest.param(dict(x_shape=[2, 2, 2], indices=[[0, 0], [0, 1]], updates=[[6.7j, 9j], [45j, 8.3j]]),
+                     marks=pytest.mark.precommit_tf_fe),
+        dict(x_shape=[2, 2, 2], indices=[[1]], updates=[[[6.7j, 9j], [45j, 8.3j]]]),
+
+    ]
```
Comment (rkazants) [inline tests/layer_tests/tensorflow_tests/test_tf_ScatterND.py:117]: please double-check instructions how to support complex tensors. Inputs of a model cannot be of complex type, we support complex tensors only between intermediate layers.
Please fix layer test

**Review by rkazants — CHANGES_REQUESTED** [review]
need to fix

File: src/frontends/tensorflow_common/src/op/scatter_nd.cpp [hunk 3]
```diff
@@ -15,15 +15,28 @@ namespace frontend {
...
     auto shape = node.get_input(2);
 
+    auto complex_type_mark_updates = as_type_ptr<ComplexTypeMark>(updates.get_node_shared_ptr());
+
+    if (complex_type_mark_updates) {
+        updates = complex_type_mark_updates->input_value(0);
+    }
+
     auto input_data = create_same_type_const<int32_t>(updates, vector<int32_t>{0}, Shape{1});
     auto broadcast = make_shared<v3::Broadcast>(input_data, shape);
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/scatter_nd.cpp:30]: `broadcast` should also have `2` auxiliary dimension so `shape` needs to be updated.
Please implement complex type and regular type separately, i.e. if (complex_type_mark_updates) {impl...} else {existing logic}

**Comment by MonalSD** [comment]
@rkazants addressed the changes. Please do give a look

File: src/frontends/tensorflow_common/src/op/scatter_nd.cpp [hunk 4]
```diff
@@ -15,15 +15,33 @@ namespace frontend {
...
+    auto broadcast = make_shared<v3::Broadcast>(input_data, updated_shape);
+
     auto scatter_nd = make_shared<v3::ScatterNDUpdate>(broadcast, input_indices, updates);
     set_node_name(node.get_name(), scatter_nd);
+
+    if (complex_type_inputs) {
+        auto complex_scatter_nd =
+            make_shared<ComplexTypeMark>(scatter_nd, complex_type_mark_updates->get_complex_part_type());
+        return {complex_scatter_nd};
+    }
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/scatter_nd.cpp:43]: you need to add auxiliary dimension size `2` only for complex type. Now you are adding to all cases. Please put complex type under `if` branch.
Also in order to update `shape` with `2`, please do the following:
1. create constant with value [2] of the same type as shape. Use helper `create_same_type_const`
2. update `shape` using Concat operation
After that you can call `Broadcast` etc.

Comment (MonalSD) [reply src/frontends/tensorflow_common/src/op/scatter_nd.cpp]: @rkazants addressed the changes. Please give it a look

**Comment by MonalSD** [comment]
@rkazants please give your inputs to proceed.

File: src/frontends/tensorflow_common/src/op/scatter_nd.cpp [hunk 5]
```diff
@@ -15,10 +17,30 @@ namespace frontend {
...
     auto shape = node.get_input(2);
+    auto complex_type_mark_updates = as_type_ptr<ComplexTypeMark>(updates.get_node_shared_ptr());
+
+    if (complex_type_mark_updates) {
+        updates = complex_type_mark_updates->input_value(0);
+        // Add two auxiliary dimensions to the shape tensor
+        Shape shape_dims = shape.get_shape();
+        shape_dims.insert(shape_dims.begin(), 2);
+        auto aux_shape = create_same_type_const<int32_t>(shape, std::vector<int32_t>{2}, Shape{1});
+        auto updated_shape = make_shared<v0::Concat>(OutputVector{aux_shape, shape}, 0);
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/scatter_nd.cpp:32]: please construct a graph using `ShapeOf` and `Concat` operations to append auxiliary dimension

File: src/frontends/tensorflow_common/src/op/scatter_nd.cpp [hunk 6]
```diff
@@ -15,10 +17,30 @@ namespace frontend {
...
+
+        auto input_data = create_same_type_const<int32_t>(updates, vector<int32_t>{0}, Shape{1});
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/scatter_nd.cpp:34]: please create a scalar = 0 of the same type as `updates`

**Comment by MonalSD** [comment]
@rkazants addressed the changes please give it a look.

**Comment by MonalSD** [comment]
@rkazants please review the changes

**Pull request closed without merging**